controller/tests/update_test_helpers.ml

```diff
@@ -0,0 +1,366 @@
+    Format.sprintf
+            "booted=%s\tprimary=%s\tvsns%s"
+            (slot_to_string booted_slot)
+            (Option.map slot_to_string primary_slot |> Option.value ~default:"-")
+            (version_info_to_string input_versions)
+
+(** Tests if the input UpdateService run with the given [system_slot_spec]
+    [case] scenario produces the expected outcome state (defined by
+    [slot_spec_to_outcome] and [state_matches_expected_outcome]).
+
+    This is used to test that all possible booted/primary and version
+    combinations lead to the correct install/no-install action.
+*)
+let test_slot_spec_combo_case case =
+    let expected_outcome = slot_spec_to_outcome case in
+    let expected_outcome_str =
+            (sexp_of_expected_outcomes expected_outcome |> Sexplib.Sexp.to_string)
+    in
+    let test_case_descr =
+        Format.sprintf
+            "%s\t->\t%s"
+            (slot_spec_to_string case)
+            expected_outcome_str
+    in
+    Alcotest_lwt.test_case test_case_descr `Quick (fun _ () ->
+        let mocks = init_test_deps () in
+
+        let () = setup_mocks_from_system_slot_spec mocks case in
+
+        let module UpdateServiceI = (val mocks.update_service) in
+        let%lwt out_state = UpdateServiceI.run_step GettingVersionInfo in
+        if state_matches_expected_outcome out_state expected_outcome then
+            Lwt.return ()
+        else
+            Alcotest.fail (Format.sprintf
+                "Reached state [%s] does not match expected outcome [%s]"
+                (statefmt out_state)
+                expected_outcome_str
+            )
+    )
+
+
+(* === Update scenario spec framework === *)
+
+type action_descr = string
+type action_check = Update.state -> bool Lwt.t
+type mock_update = unit -> unit
+
+type scenario_spec =
+  | StateReached of Update.state
+  | ActionDone of action_descr * action_check
+  | UpdateMock of mock_update
+
+let specfmt spec = match spec with
+    | StateReached s -> "StateReached: " ^ (statefmt s);
+    | ActionDone (descr, c) -> "ActionDone: " ^ descr;
+    | UpdateMock _ -> "UpdateMock: <fun>"
+
+let _MAGIC_PAT = "<..>"
+
+(* string equality, but the magic patern `<..>` is treated
+  as a placeholder for any sub-string. The implementation converts the
+  `expected` string to a regex where the magic pattern is replaced with ".*",
+  while being careful to `Str.quote` the rest of the string to not accidentally
+  treat them as regex expressions.
+*)
+let str_match_with_magic_pat expected actual =
+    let open Str in
+    let magic_pattern = regexp_string _MAGIC_PAT in
+    let exp_parts = full_split magic_pattern expected in
+    let exp_regexp = regexp @@ String.concat "" @@ List.map (fun (p) ->
+        match p with
+            | Text a -> quote a
+            | Delim _ -> ".*"
+    ) exp_parts in
+    string_match exp_regexp actual 0
+
+
+let state_formatter out inp = Format.fprintf out "%s" (statefmt inp)
+
+let t_state =
```

Comment:
The `t` stands for testable, because this defines an instance of `Alcotest.testable` for `state`. I could rename it to `testable_state` if that's helpful.